### PR TITLE
Rename SharedMutex and introduce Mutex

### PR DESCRIPTION
**Pull request by jac3km4**
I was looking into the RTTI and there appear to be 4 types of locks:
- spin lock (currently named SpinLock in the SDK)
- shared spin lock (currently named SharedMutex in the SDK)
- mutex (uses critical section, does not exist in the SDK)
- shared mutex (uses srwlock, does not exist in the SDK)

In this PR I do a little bit of cleanup:
- rename `SharedMutex` to `SharedSpinLock`, because that's the more appopriate name, in c++ a shared mutex corresponds to a srwlock on Windows, not a spin lock
- introduce `Mutex`
- update the RTTI system to make use of `Mutex` and `SharedSpinLock`

I didn't add the rwlock because I don't think it's needed anywhere in the SDK right now.

### CODE REVIEW

**Review by psiberx — APPROVED** [review]

File: include/RED4ext/Mutex.hpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    void Lock();
+    void Unlock();
+
+    // --------------------------------------------
+    // -- support for lock_guard and shared_lock --
+    // --------------------------------------------
+
+    void lock();
+    void unlock();
+};
```
Comment (WopsS) [inline include/RED4ext/Mutex.hpp:27]:
```suggestion
struct Mutex
{
    Mutex();
    Mutex(const Mutex&) = delete;
    Mutex(Mutex&&) = delete;
    Mutex& operator=(const Mutex&) = delete;
    Mutex& operator=(Mutex&&) = delete;

    void Lock();
    void Unlock();

    // --------------------------------------------
    // -- support for lock_guard and shared_lock --
    // --------------------------------------------

    void lock();
    void unlock();
private:
    CRITICAL_SECTION m_cs;
};
```

File: include/RED4ext/SharedSpinLock.hpp [hunk 2]
```diff
@@ -4,18 +4,18 @@
 
 namespace RED4ext
 {
-struct SharedMutex
+struct SharedSpinLock
```
Comment (WopsS) [inline include/RED4ext/SharedSpinLock.hpp:7]: Imo, we should add an alias to this and mark `SharedMutex` as deprecated.

**Review by WopsS — CHANGES_REQUESTED** [review]
Thanks! I added my comments.

File: include/RED4ext/SharedSpinLock.hpp (hunk 2 above)
Comment (jac3km4) [reply include/RED4ext/SharedSpinLock.hpp:7]: I've added an alias with deprecation

**Review by WopsS — APPROVED** [review]

**Pull request merged**